File: 414.third-maximum-number.py
#
# @lc app=leetcode id=414 lang=python3
#
# [414] Third Maximum Number
#

# @lc code=start
class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        max_number = None
        second_maximum_number = None
        third_maximum_number = None
        for num in nums:
            if num in [max_number,second_maximum_number,third_maximum_number]:
                continue
            if max_number is None:
                max_number = num
                continue
            if num > max_number:
                max_number,num = num,max_number
            if second_maximum_number is None:
                second_maximum_number = num
                continue
            if num > second_maximum_number:
                second_maximum_number,num = num,second_maximum_number
            if third_maximum_number is None:
                third_maximum_number = num
                continue
            if  num > third_maximum_number:
                third_maximum_number,num = num,third_maximum_number
        return third_maximum_number if third_maximum_number is not None else max_number
            


# @lc code=end


if __name__ == "__main__":
    pass

File: golang/414.third-maximum-number.go
/*
 * @lc app=leetcode.cn id=414 lang=golang
 *
 * [414] Third Maximum Number
 */
package main

import (
	"container/heap"
)

// @lc code=start

type IntHeap []int

func (h IntHeap) Len() int { return len(h) }

// 重新定义 less
func (h IntHeap) Less(i, j int) bool { return h[i] > h[j] }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x interface{}) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents.
	*h = append(*h, x.(int))
}

func (h *IntHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func thirdMax(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	}
	if len(nums) == 2 {
		if nums[1] > nums[0] {
			return nums[1]
		} else {
			return nums[0]
		}
	}

	h := IntHeap(nums)
	heap.Init(&h)

	largest := heap.Pop(&h).(int)
	second := int(0)
	third := second

	second = heap.Pop(&h).(int)
	for second == largest {
		if h.Len() == 0 {
			return largest
		}
		second = heap.Pop(&h).(int)
	}

	if h.Len() == 0 {
		return largest
	}
	third = heap.Pop(&h).(int)
	for third == second {
		if h.Len() == 0 {
			return largest
		}
		third = heap.Pop(&h).(int)

	}
	return third
}

// @lc code=end

func main() {
	thirdMax([]int{2, 2, 3, 1})
}
